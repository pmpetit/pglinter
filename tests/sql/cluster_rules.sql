-- Test for C001 and C002 cluster rules
BEGIN;

DROP EXTENSION IF EXISTS pglinter CASCADE;

CREATE EXTENSION IF NOT EXISTS pglinter;

-- Test cluster rules
SELECT 'Testing cluster rules C001 and C002...' as test_info;

-- C001: Memory configuration issues
-- This rule checks max_connections and work_mem settings
-- It will analyze current PostgreSQL configuration
SELECT 'Current PostgreSQL configuration:' as info;
SELECT current_setting('max_connections') as max_connections;
SELECT current_setting('work_mem') as work_mem;

-- C002: Insecure pg_hba.conf entries
-- This rule checks for insecure authentication methods
-- Note: This may not trigger in test environment

-- Run cluster check
SELECT pglinter.perform_cluster_check();

-- Test rule management for cluster rules
SELECT pglinter.explain_rule('C001');
SELECT pglinter.explain_rule('C002');

-- Test enabling/disabling cluster rules
SELECT pglinter.is_rule_enabled('C001') AS c001_enabled;
SELECT pglinter.is_rule_enabled('C002') AS c002_enabled;

-- Disable C001 temporarily
SELECT pglinter.disable_rule('C001') AS c001_disabled;
SELECT pglinter.perform_cluster_check(); -- Should skip C001

-- Re-enable C001
SELECT pglinter.enable_rule('C001') AS c001_reenabled;
SELECT pglinter.perform_cluster_check(); -- Should include C001 again

-- Test output to file for cluster rules
SELECT pglinter.perform_cluster_check('/tmp/cluster_test.sarif');

-- Show all cluster-related information
SELECT 'Cluster configuration summary:' as info;
SELECT
    current_setting('max_connections') as max_connections,
    current_setting('work_mem') as work_mem,
    current_setting('shared_buffers') as shared_buffers,
    current_setting('effective_cache_size') as effective_cache_size;

ROLLBACK;
