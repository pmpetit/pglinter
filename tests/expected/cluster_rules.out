-- Test for C001 and C002 cluster rules
BEGIN;
DROP EXTENSION IF EXISTS pglinter CASCADE;
NOTICE:  extension "pglinter" does not exist, skipping
CREATE EXTENSION IF NOT EXISTS pglinter;
-- Test cluster rules
SELECT 'Testing cluster rules C001 and C002...' as test_info;
               test_info                
----------------------------------------
 Testing cluster rules C001 and C002...
(1 row)

-- C001: Memory configuration issues
-- This rule checks max_connections and work_mem settings
-- It will analyze current PostgreSQL configuration
SELECT 'Current PostgreSQL configuration:' as info;
               info                
-----------------------------------
 Current PostgreSQL configuration:
(1 row)

SELECT current_setting('max_connections') as max_connections;
 max_connections 
-----------------
 100
(1 row)

SELECT current_setting('work_mem') as work_mem;
 work_mem 
----------
 4MB
(1 row)

-- C002: Insecure pg_hba.conf entries
-- This rule checks for insecure authentication methods
-- Note: This may not trigger in test environment
-- Run cluster check
SELECT pglinter.perform_cluster_check();
NOTICE:  üîç pglinter found 1 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ÑπÔ∏è  [C002] INFO: Please manually check pg_hba.conf for insecure trust/password methods
NOTICE:  ==================================================
NOTICE:  üìä Summary: 0 error(s), 0 warning(s), 1 info
NOTICE:  üü¢ Only informational messages - good job!
 perform_cluster_check 
-----------------------
 t
(1 row)

-- Test rule management for cluster rules
SELECT pglinter.explain_rule('C001');
NOTICE:  üìñ Rule Explanation for C001
============================================================

üéØ Rule Name: MaxConnectionsByWorkMemIsNotLargerThanMemory
üìã Scope: CLUSTER

üìù Description:
Number of cx (max_connections * work_mem) is not greater than memory.

‚ö†Ô∏è  Message Template:
work_mem * max_connections is bigger than ram.

üîß How to Fix:
   1. downsize max_connections or upsize memory
============================================================
 explain_rule 
--------------
 t
(1 row)

SELECT pglinter.explain_rule('C002');
NOTICE:  üìñ Rule Explanation for C002
============================================================

üéØ Rule Name: PgHbaEntriesWithMethodTrustOrPasswordShouldNotExists
üìã Scope: CLUSTER

üìù Description:
This configuration is extremely insecure and should only be used in a controlled, non-production environment for testing purposes. In a production environment, you should use more secure authentication methods such as md5, scram-sha-256, or cert, and restrict access to trusted IP addresses only.

‚ö†Ô∏è  Message Template:
{0} entries in pg_hba.conf with trust or password authentication method exceed the warning threshold: {1}.

üîß How to Fix:
   1. change trust or password method in pg_hba.conf
============================================================
 explain_rule 
--------------
 t
(1 row)

-- Test enabling/disabling cluster rules
SELECT pglinter.is_rule_enabled('C001') AS c001_enabled;
 c001_enabled 
--------------
 t
(1 row)

SELECT pglinter.is_rule_enabled('C002') AS c002_enabled;
 c002_enabled 
--------------
 t
(1 row)

-- Disable C001 temporarily
SELECT pglinter.disable_rule('C001') AS c001_disabled;
NOTICE:  üî¥ Rule C001 has been disabled
 c001_disabled 
---------------
 t
(1 row)

SELECT pglinter.perform_cluster_check(); -- Should skip C001
NOTICE:  üîç pglinter found 1 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ÑπÔ∏è  [C002] INFO: Please manually check pg_hba.conf for insecure trust/password methods
NOTICE:  ==================================================
NOTICE:  üìä Summary: 0 error(s), 0 warning(s), 1 info
NOTICE:  üü¢ Only informational messages - good job!
 perform_cluster_check 
-----------------------
 t
(1 row)

-- Re-enable C001
SELECT pglinter.enable_rule('C001') AS c001_reenabled;
NOTICE:  ‚úÖ Rule C001 has been enabled
 c001_reenabled 
----------------
 t
(1 row)

SELECT pglinter.perform_cluster_check(); -- Should include C001 again
NOTICE:  üîç pglinter found 1 issue(s):
NOTICE:  ==================================================
NOTICE:  ‚ÑπÔ∏è  [C002] INFO: Please manually check pg_hba.conf for insecure trust/password methods
NOTICE:  ==================================================
NOTICE:  üìä Summary: 0 error(s), 0 warning(s), 1 info
NOTICE:  üü¢ Only informational messages - good job!
 perform_cluster_check 
-----------------------
 t
(1 row)

-- Test output to file for cluster rules
SELECT pglinter.perform_cluster_check('/tmp/cluster_test.sarif');
 perform_cluster_check 
-----------------------
 t
(1 row)

-- Show all cluster-related information
SELECT 'Cluster configuration summary:' as info;
              info              
--------------------------------
 Cluster configuration summary:
(1 row)

SELECT
    current_setting('max_connections') as max_connections,
    current_setting('work_mem') as work_mem,
    current_setting('shared_buffers') as shared_buffers,
    current_setting('effective_cache_size') as effective_cache_size;
 max_connections | work_mem | shared_buffers | effective_cache_size 
-----------------+----------+----------------+----------------------
 100             | 4MB      | 128MB          | 4GB
(1 row)

ROLLBACK;
